Core/src/Detector/VolumeStructureBuilder.cpp

```diff
@@ -0,0 +1,176 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2023 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "Acts/Detector/VolumeStructureBuilder.hpp"
+
+#include "Acts/Detector/PortalGenerators.hpp"
+#include "Acts/Geometry/ConeVolumeBounds.hpp"
+#include "Acts/Geometry/CuboidVolumeBounds.hpp"
+#include "Acts/Geometry/CutoutCylinderVolumeBounds.hpp"
+#include "Acts/Geometry/CylinderVolumeBounds.hpp"
+#include "Acts/Geometry/GenericCuboidVolumeBounds.hpp"
+#include "Acts/Geometry/TrapezoidVolumeBounds.hpp"
+#include "Acts/Utilities/Enumerate.hpp"
+#include "Acts/Utilities/Helpers.hpp"
+
+Acts::Experimental::VolumeStructureBuilder::VolumeStructureBuilder(
+    const Acts::Experimental::VolumeStructureBuilder::Config& cfg,
+    std::unique_ptr<const Acts::Logger> logger)
+    : IExternalStructureBuilder(), m_cfg(cfg), m_logger(std::move(logger)) {
+  // Sanity cross-checks
+  if (m_cfg.boundValues.empty() and not m_cfg.extent.has_value()) {
+    throw std::invalid_argument(
+        "VolumeStructureBuilder: no extent nor boundary values given");
+  }
+  // Check for the bounds type
+  if (m_cfg.boundsType == VolumeBounds::BoundsType::eOther) {
+    throw std::invalid_argument(
+        "VolumeStructureBuilder: no known volume bounds type provided.");
+  }
+}
+
+Acts::Experimental::ExternalStructure
+Acts::Experimental::VolumeStructureBuilder::construct(
+    [[maybe_unused]] const Acts::GeometryContext& gctx) const {
+  // The volume bounds to be constructed
+  std::unique_ptr<VolumeBounds> volumeBounds = nullptr;
+
+  // A transform
+  auto transform = Transform3::Identity();
+  std::vector<ActsScalar> boundValues = m_cfg.boundValues;
+
+  // This code dispatches into the dedicated volume types
+  switch (m_cfg.boundsType) {
+    case VolumeBounds::BoundsType::eCone: {
+      ACTS_VERBOSE("Building conical volume bounds.");
+      // Cone translation - only pre-defined values
+      if (boundValues.size() < 5u) {
```

Comment:
Maybe the minimum size of boundValues for each BoundsType could be saved somewhere within the class?

Reply:
That's effectively already in place with `eSize`.

https://github.com/acts-project/acts/blob/55b07bc69d9643861b6f8825656ae1208147fb41/Core/include/Acts/Geometry/CylinderVolumeBounds.hpp#L74-L83

But it's a convention, not a guarantee.

Reply:
Thing is, you can "guess" the later one, at leas fill them with default values, which I would like to keep.